### DISCUSSION TITLE
Modernising the RetroBSD codebase

### DISCUSSION THREAD

**Discussion by majenkotech** [RetroBSD/retrobsd#83] [General]
The bit rot has set in, as is to be expected. It's getting harder and harder to compile the old code that makes up RetroBSD, so we really need to modernise it and rewrite / update portions of it to allow it to be compiled with a modern compiler.

I guess step one is to select a reference compiler to work with...

I have a vanilla MIPS GCC compiler that I have built to work on as a replacement for `pic32-gcc` from the chipKIT project. It doesn't have any PIC32 stuff in it yet, but for vanilla MIPS it works fine. It's GCC 10.2.0:

* https://github.com/MajenkoProjects/chipkit-compiler/releases/tag/v0.0.1

Serge has mentioned that there may be a Clang compiler for PIC32, but I have no knowledge or experience of that.

From an initial test using gcc 10.2.0 I have found some things that need work:

1. Kernel header files are full of variables. These need removing and replacing with `extern` references to the variables defined in source files instead
2. There are large amounts of `register` variable definitions. The generally accepted global opinion is that `register` is pointless and the compiler now does a better job of optimizing than humans do. Indeed, some compilers ignore the hint and only enforce the "you can't get an address of a register" side effect
3. The code is a bit of a mess - layout wise. There's plenty of errors about `if` statements not covering things you may think it does since the indenting is wrong. Easily fixed by feeding it all through `astyle`, but better to review them all in case there's any errors there that should be corrected
4. There's a number of other warnings and errors about incorrect printf formats, some incorrect pointers and function types as parameters, etc. 
5. It's riddled with the old style format of function prototypes where the parameter types are listed separate from the function itself, which is not only confusing, but encourages "implied" variable types which modern compilers do not like (and neither do I). They should IMHO all be changed to proper modern "normal" function formats.

So plenty to review and fix just in the kernel.

**Comment by sergev** [comment]
I'm getting positive results with Clang and LLVM linker. I was able to build a simple pic32 example:
> /usr/lib/llvm-12/bin/**clang -target mipsel -mcpu=mips32r2 -mabi=o32 -mfloat-abi=soft** -Os -g -Wall -Werror -DPIC32MX7 -nostdlib -fomit-frame-pointer -finline-hint-functions -c uart.c
/usr/lib/llvm-12/bin/**ld.lld -m elf32ltsmip** -T using-bootloader.ld -e _start -g uart.o -o uart.elf
/usr/lib/llvm-12/bin/llvm-objcopy -O ihex uart.elf uart.hex
/usr/lib/llvm-12/bin/llvm-objdump --mcpu=mips32r2 -d -S uart.elf > uart.dis

No need for extra toolchain. Regular clang and lld packages do just fine:
> sudo apt install clang-12 lld-12

Other clang version should work as well, I guess. I chose version 12 as I have it installed both on my Ubuntu and MacOS machines.

Only minor changes to the linker script are required: https://github.com/sergev/vak-opensource/commit/5d070c16be9c92e499e153b9a998b22ba2032c85#diff-95c337c9ca2f033f5bdb95e71de2c58aa0d2151268b870b0b72b097cb9fa2308

**Reply by majenkotech** [reply]
llvm 13 which Manjaro has by default (that's my favourite distro these days...) seems to work fine too, though I haven't tried with real code yet...

**Comment by sergev** [comment]
I created a branch [clang](https://github.com/RetroBSD/retrobsd/tree/clang), and started modifying build scripts for clang, and fixing detected issues in the sources. 

I found another big problem: the assembler in RetroBSD is not compatible with clang. When compiling libraries in RetroBSD native format, we invoke clang to convert C code to assembly, and then we use `retrobsd/src/cmd/as/as.c` to create object files in aout format. There are tho problems:
* Our `as` does not understand all the command line options clang passes to it.
* Clang generates a lot of additional assembly statements which our assembler cannot recognize, like `.abicalls` and many others.

Anyway, I'll try to move forward. Not sure how much effort it would take.

**Reply by majenkotech** [reply]
I guess long-term we want to upgrade retrobsd-as to handle clang-generated assembly, but for a stop-gap would it be simpler to write a filter script that strips out the unwanted stuff and converts any special instructions / macros to what retrobsd-as expects?

**Reply by sergev** [reply]
Definitely possible. But updating the assembler is not much complicated than writing a filter script. For now, I'm just going to disable this part of the build. No native libraries.

**Reply by calmsacibis995** [reply]
Another thing that needs modernizing is the QEMU emulator. Serge's fork of QEMU is quite outdated and can only be built on older versions of Ubuntu. (I've only tested 14.04, it should also work on 16.04) This means that someone who wants to run RetroBSD (assuming they somehow overcame the bit rot) would also need to update Serge's fork of QEMU, since QEMU has no PIC32 support.
